pkg/object/scaffold.go

```diff
@@ -0,0 +1,105 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package object
+
+import (
+	"io"
+	"strings"
+	"text/template"
+)
+
+var (
+	typesTemplateRaw = `// {{.Resource.Kind}}Spec defines the desired state of {{.Resource.Kind}}
+type {{.Resource.Kind}}Spec struct {
+	// INSERT ADDITIONAL SPEC FIELDS -- desired state of cluster
+{{- if .AdditionalHelp }}
+{{- range .AdditionalHelp | SplitLines }}
+	// {{.}}
+{{- end }}
+{{- end }}
+}
+
+// {{.Resource.Kind}}Status defines the observed state of {{.Resource.Kind}}.
+// It should always be reconstructable from the state of the cluster and/or outside world.
+type {{.Resource.Kind}}Status struct {
+	// INSERT ADDITIONAL STATUS FIELDS -- observed state of cluster
+{{- if .AdditionalHelp }}
+{{- range .AdditionalHelp | SplitLines }}
+	// {{.}}
+{{- end }}
+{{- end }}
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+{{- if .GenerateClients }}
+// +genclient
+{{- if not .Resource.Namespaced }}
+// +genclient:nonNamespaced
+{{- end }}
+{{- end }}
+
+// {{.Resource.Kind}} is the Schema for the {{ .Resource.Resource }} API
+// +k8s:openapi-gen=true
+type {{.Resource.Kind}} struct {
+	metav1.TypeMeta   ` + "`" + `json:",inline"` + "`" + `
+	metav1.ObjectMeta ` + "`" + `json:"metadata,omitempty"` + "`" + `
+
+	Spec   {{.Resource.Kind}}Spec   ` + "`" + `json:"spec,omitempty"` + "`" + `
+	Status {{.Resource.Kind}}Status ` + "`" + `json:"status,omitempty"` + "`" + `
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+{{- if and (.GenerateClients) (not .Resource.Namespaced) }}
+// +genclient:nonNamespaced
+{{- end }}
+
+// {{.Resource.Kind}}List contains a list of {{.Resource.Kind}}
+type {{.Resource.Kind}}List struct {
+	metav1.TypeMeta ` + "`" + `json:",inline"` + "`" + `
+	metav1.ListMeta ` + "`" + `json:"metadata,omitempty"` + "`" + `
+	Items           []{{ .Resource.Kind }} ` + "`" + `json:"items"` + "`" + `
+}
+`
+	typesTemplateHelpers = template.FuncMap{
+		"SplitLines": func(raw string) []string { return strings.Split(raw, "\n") },
+	}
+
+	typesTemplate = template.Must(template.New("object-scaffolding").Funcs(typesTemplateHelpers).Parse(typesTemplateRaw))
+)
+
+// ScaffoldOptions describes how to scaffold out a Kubernetes object
+// with the basic metadata and comment annotations required to generate code
+// for and conform to runtime.Object and metav1.Object.
+type ScaffoldOptions struct {
+	Resource        Resource
+	AdditionalHelp  string
+	GenerateClients bool
+}
+
+// Validate validates the options, returning an error if anything is invalid.
+func (o *ScaffoldOptions) Validate() error {
+	if err := o.Resource.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Scaffold prints the Kubernetes object scaffolding to the given output.
+func (o *ScaffoldOptions) Scaffold(out io.Writer) error {
+	return typesTemplate.Execute(out, o)
+}
```

Comment:
Initially we had scaffolding tools under controller-tools and fixing bugs required multiple steps across two repos which is:
 
 - fixing the bug in controller-tools
 - tagging the controller-tools.
 - then, updating the deps in kubebuilder

Which felt a bit slow. Since we didn't have any users for scaffold tools in controller-tools, we moved the scaffold tools to kubebuilder repo. It simplified the dev/test/release workflow. But then we were iterating faster because KB was in early days, we may not feel the same pain now because our requirements are matured now.